src/components/clusterexplorer/explorer.ts

```diff
@@ -0,0 +1,123 @@
+                this.refresh();
+            }
+        });
+    }
+
+    getTreeItem(element: ClusterExplorerNode): vscode.TreeItem | Thenable<vscode.TreeItem> {
+        const treeItem = element.getTreeItem();
+
+        const treeItem2 = providerResult.transform(treeItem, (ti) => {
+            if (ti.collapsibleState === vscode.TreeItemCollapsibleState.None && this.extenders.some((e) => e.contributesChildren(element))) {
+                ti.collapsibleState = vscode.TreeItemCollapsibleState.Collapsed;
+            }
+        });
+
+        let treeItem3 = treeItem2;
+        for (const c of this.customisers) {
+            treeItem3 = providerResult.transformPossiblyAsync(treeItem2, (ti) => c.customize(element, ti));
+        }
+        return treeItem3;
+}
+
+    getChildren(parent?: ClusterExplorerNode): vscode.ProviderResult<ClusterExplorerNode[]> {
+        const baseChildren = this.getChildrenBase(parent);
+        const contributedChildren = this.extenders
+                                        .filter((e) => e.contributesChildren(parent))
+                                        .map((e) => e.getChildren(parent));
+        return providerResult.append(baseChildren, ...contributedChildren);
+    }
+
+    private getChildrenBase(parent?: ClusterExplorerNode): vscode.ProviderResult<ClusterExplorerNode[]> {
+        if (parent) {
+            return parent.getChildren(this.kubectl, this.host);
+        }
+        return this.getClusters();
+    }
+
+    refresh(): void {
+        this.onDidChangeTreeDataEmitter.fire();
+    }
+
+    registerExtender(extender: ExplorerExtender<ClusterExplorerNode>): void {
+        this.extenders.push(extender);
+        // In the case where an extender contributes at top level (sibling to cluster nodes),
+        // the tree view can populate before the extender has time to register.  So in this
+        // case we need to kick off a refresh.  But... it turns out that if we just fire the
+        // change event, VS Code goes 'oh well I'm just drawing the thing now so I'll be
+        // picking up the change, no need to repopulate a second time.'  Even with a delay
+        // there's a race condition.  But it seems that if we pipe it through the refresh
+        // *command* (as refreshExplorer does) then it seems to work... ON MY MACHINE TM anyway.
+        //
+        // This is a pretty niche case, so I'm not too worried if this isn't perfect.
+        //
+        // TODO: VS Code now doesn't require a reload on extension install.  Do we need
+        // to listen for the extension install event and refresh, in case an extension
+        // attempts to contribute while the tree view is already open?
+        //
+        // TODO: we need to check collapsibleStates in case someone adds child nodes to a
+        // parent which currently has CollapsibleState.None.
+        if (extender.contributesChildren(undefined)) {
+            sleep(50).then(() => refreshExplorer());
```

Comment:
This is clearly a hack (and duplicated below) refactor so that we have one impl of the hack and an explanation of why we need it with that impl.